### PR TITLE
[stdlib] Portable random function

### PR DESCRIPTION
**Pull request by tinysun212**
#### Resolved bug number: N/A

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

Changed rand32() implementation using a random engine in C++11 `<random>`.

### CODE REVIEW

File: stdlib/public/stubs/LibcShims.cpp [hunk 1]
```diff
@@ -69,11 +70,19 @@ size_t _swift_stdlib_malloc_size(const void *ptr) {
...
+  return MersenneRandom();
+}
 
 __swift_uint32_t
-_swift_stdlib_arc4random_uniform(__swift_uint32_t upper_bound) {
-  return arc4random_uniform(upper_bound);
+_swift_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound) {
+  if (upper_bound > 0)
+    upper_bound--;
+  std::uniform_int_distribution<int> RandomUniform(0, upper_bound);
```
Comment (gribozavr) [inline stdlib/public/stubs/LibcShims.cpp]: Why not `std::uniform_int_distribution< __swift_uint32_t >`?

File: stdlib/private/SwiftPrivate/PRNG.swift [hunk 2]
```diff
@@ -14,19 +14,19 @@ import SwiftShims
...
+  return _swift_stdlib_cxx11_mt19937()
 }
 
 @inline(never) @_semantics("stdlib_binary_only") // Hide the libbsd dependency
 public func rand32(exclusiveUpperBound limit: UInt32) -> UInt32 {
-  return _swift_stdlib_arc4random_uniform(limit)
+  return _swift_stdlib_cxx11_mt19937_uniform(limit)
 }
 
 @inline(never) @_semantics("stdlib_binary_only") // Hide the libbsd dependency
```
Comment (gribozavr) [inline stdlib/private/SwiftPrivate/PRNG.swift]: I think you can drop `@inline(never) @_semantics("stdlib_binary_only")` from these three functions now, these attributes are not serving any purpose now that we don't depend on libbsd.

**Comment by gribozavr** [comment]
@tinysun212 LGTM, only minor comments!  Let's test this as well:

@swift-ci Please test

File: validation-test/stdlib/Set.swift [hunk 3]
```diff
@@ -105,7 +105,7 @@ func helperDeleteThree(k1: TestKeyTy, _ k2: TestKeyTy, _ k3: TestKeyTy) {
 }
 
 func uniformRandom(max: Int) -> Int {
-  return Int(arc4random_uniform(UInt32(max)))
+  return Int(cxx11_mt19937_uniform(UInt32(max)))
```
Comment (gribozavr) [inline validation-test/stdlib/Set.swift]: Looks like this should be calling `rand32()` instead.

**Comment by gribozavr** [comment]
LGTM, let's merge.

@swift-ci Please test and merge

**Comment by gribozavr** [comment]
@tinysun212 Sorry!  CI hit an error:

```
swift/validation-test/stdlib/Set.swift:108:21: error: missing argument label 'exclusiveUpperBound:' in call
  return Int(rand32(UInt32(max)))
                    ^
                    exclusiveUpperBound: 
```

**Comment by tinysun212** [comment]
Fixed and squashed.

**Comment by gribozavr** [comment]
@swift-ci Please test and merge

**Pull request merged**